### ISSUE TITLE
[🐛 bug report] Wrong indexing order when converting Mueller matrices into NumPy / Torch arrays

### ISSUE THREAD

**Issue by class255** [mitsuba-renderer/mitsuba3#996]
## Summary

Converting Mueller matrices into NumPy/PyTorch array/tensor yields wrong entry orders.

## System configuration

System information:

  OS: Windows 11
  CPU: 12th Gen Intel(R) Core(TM) i9-12900K 3.19 GHz
  GPU: NVIDIA GeForce RTX 4090
  Python version: 3.11
  LLVM version: ...
  CUDA version: 12.0
  NVidia driver: 528.02

  Dr.Jit version: 0.4.2
  Mitsuba version: 3.3.0
     Compiled with: Manually compiled with CMake & Visual Studio
     Variants compiled: ['scalar_rgb', 'scalar_spectral', 'cuda_ad_rgb', 'llvm_ad_rgb', 'scalar_rgb_polarized', 'scalar_spectral_polarized', 'cuda_rgb', 'cuda_spectral', 'cuda_rgb_polarized', 'cuda_spectral_polarized', 'llvm_rgb', 'llvm_spectral', 'llvm_rgb_polarized', 'llvm_spectral_polarized', 'cuda_ad_rgb_polarized', 'cuda_ad_spectral_polarized', 'cuda_ad_spectral_polarized_double', 'llvm_ad_rgb_polarized', 'llvm_ad_spectral_polarized', 'llvm_ad_spectral_polarized_double']

  NumPy version: 1.24.3
  PyTorch version: 2.1.1+cu118

## Description

Code:
```python
import numpy as np
import torch
import drjit as dr
import mitsuba as mi
mi.set_variant('cuda_rgb_polarized')
np.set_printoptions(precision=8, linewidth=100)

# ---------- Initialize variables ----------
def bsdf_pplastic(diffuse_reflectance, specular_reflectance, ior, alpha) -> mi.BSDF:
    bsdf_dict = {'type':'pplastic',
                    'diffuse_reflectance':{
                        'type': 'rgb',
                        'value': diffuse_reflectance},
                    'specular_reflectance':{
                        'type': 'rgb',
                        'value': specular_reflectance},
                    'int_ior': ior,
                    'ext_ior': 1.0,
                    'alpha': alpha}
    return mi.load_dict(bsdf_dict)

bsdf = bsdf_pplastic([1., 0., 0.], [1., 0., 0.], 1.5, 1.0) # Use red channel only

wi = mi.Vector3f(np.array([[0.1, 0., 0.9], [0., 0.2, 0.9]]))
wo = mi.Vector3f(np.array([[0.1, 0., 0.9], [0., 0.2, 0.9]]))

# ---------- Evaluate the BSDF ----------
si = mi.SurfaceInteraction3f()
si.wi = wi
ctx = mi.BSDFContext()

res_mi = bsdf.eval(ctx, si, wo1)
print(f"# 1. Print res_mi: {type(res_mi)}")
print(f"{dr.shape(res_mi) = }")
print(res_mi)

# ---------- Convert into NumPy / PyTorch ----------
res_np = res_mi.numpy()
res_tc = res_mi.torch()
print("\n # 2. Print res_np (NumPy), which has the same result as torch")
print(f"{res_np.shape = }")
print(f"{res_tc.shape = }")
print(f"{np.allclose(res_np, torch.Tensor.cpu(res_tc).numpy()) = }")
print(res_np)

# ---------- Fix NumPy array order ----------
temp = res_np.transpose(0, 1, 3, 2) # [2, 3, 4, 4]
temp.shape = (2, 4, 4, 3)
res_np_fix = temp.transpose(0, 3, 2, 1)
print("\n # 3. Fix Numpy array order")
print(f"{res_np_fix.shape = }")
print(res_np_fix)
```

Printed results:
```python
# 1. Print res_mi: <class 'mitsuba.cuda_rgb_polarized.Spectrum'>
dr.shape(res_mi) = [4, 4, 3, 2]
[[[[0.26722630858421326, 0.0022400030866265297, 1.0517184900749044e-09, 0.0],
   [0.0022399723529815674, 0.0036822401452809572, 1.2878649535696951e-11, 0.0],
   [0.0, 0.0, -0.0035062560345977545, 0.0],
   [0.0, 0.0, 0.0, -0.0035062560345977545]],
  [[0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0]],
  [[0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0]]],
 [[[0.2672477662563324, 0.0038737845607101917, -0.0008530994527973235, 0.0],
   [-0.002602330408990383, 0.003520220285281539, 1.00232628028607e-05, 0.0],
   [0.000868420407641679, 0.00016271146887447685, -0.003606732003390789, 0.0],
   [0.0, 0.0, 0.0, -0.0035044029355049133]],
  [[0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0]],
  [[0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0],
   [0.0, 0.0, 0.0, 0.0]]]]

 # 2. Print res_np (NumPy), which has the same result as torch
res_np.shape = (2, 3, 4, 4)
res_tc.shape = torch.Size([2, 3, 4, 4])
np.allclose(res_np, torch.Tensor.cpu(res_tc).numpy()) = True
[[[[ 2.67226309e-01  0.00000000e+00  0.00000000e+00  2.24000309e-03]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 2.23997235e-03  0.00000000e+00  0.00000000e+00  3.68224015e-03]]

  [[ 0.00000000e+00  0.00000000e+00  1.05171849e-09  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.50625603e-03]
   [ 0.00000000e+00  0.00000000e+00  1.28786495e-11  0.00000000e+00]]

  [[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.50625603e-03]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]]


 [[[ 2.67247766e-01  0.00000000e+00  0.00000000e+00  3.87378456e-03]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  8.68420408e-04  0.00000000e+00  0.00000000e+00]
   [-2.60233041e-03  0.00000000e+00  0.00000000e+00  3.52022029e-03]]

  [[ 0.00000000e+00  0.00000000e+00 -8.53099453e-04  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 1.62711469e-04  0.00000000e+00  0.00000000e+00 -3.60673200e-03]
   [ 0.00000000e+00  0.00000000e+00  1.00232628e-05  0.00000000e+00]]

  [[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.50440294e-03]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]]]

 # 3. Fix Numpy array order
res_np_fix.shape = (2, 3, 4, 4)
[[[[ 2.67226309e-01  2.24000309e-03  1.05171849e-09  0.00000000e+00]
   [ 2.23997235e-03  3.68224015e-03  1.28786495e-11  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00 -3.50625603e-03  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.50625603e-03]]

  [[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]

  [[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]]


 [[[ 2.67247766e-01  3.87378456e-03 -8.53099453e-04  0.00000000e+00]
   [-2.60233041e-03  3.52022029e-03  1.00232628e-05  0.00000000e+00]
   [ 8.68420408e-04  1.62711469e-04 -3.60673200e-03  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.50440294e-03]]

  [[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]

  [[ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]
   [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00]]]]
```

Sorry for long code & results.

First, to see index order intuitively, I made a pplastic `mi.BSDF` instance which has the only nonzero reflectance for red channels (B=G=0). The resulting pBSDF evaluated two pairs of directions has been stored in the variable `res_mi`.
Then we can see from the printed result with label "# 1. Print res_mi: <class 'mitsuba.cuda_rgb_polarized.Spectrum'>" that `res_mi[i,j,0,k]` can be nonzero but `res_mi[i,j,1,k]` and `res_mi[i,j,2,k]` are zeros for any i,j,k indices.

However, converting it into NumPy array `res_np = res_mi.numpy()` gives strange values.
We can see from the printed result with label "# 2. Print res_np (NumPy), which has the same result as torch" that nonzero entries distribute unexpectedly.
Converting to `.torch()` rather than `.numpy()` also produces exactly same result.
(See "np.allclose(res_np, torch.Tensor.cpu(res_tc).numpy()) = True" above)

To get the correct order of array, I needed to perform following steps (which is also contained in the above code block)
```python
temp = res_np.transpose(0, 1, 3, 2)
temp.shape = (2, 4, 4, 3)    # This `2` will change depending on the number of vectorized data
res_np_fix = temp.transpose(0, 3, 2, 1)
```

**Comment by njroussel** [comment]
Fixed in https://github.com/mitsuba-renderer/drjit/commit/4e9b718a7c3012c61b51196afde960e0c0aa160d.

You should be able to incorporate this fix locally, it's a one line change.

Thank you for reporting this issue!

**Comment by class255** [comment]
Thanks a lot!
